### ISSUE TITLE
[BUG] <Attention mask while training>

### ISSUE THREAD

**Issue by diuzi** [THUDM/ChatGLM2-6B#185]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

喂进`torch.nn.functional.scaled_dot_product_attention`的attn_mask padding位置处关注了整个序列。

### Expected Behavior

Attention mask 的实现在训练时貌似不符合预期. 假设训练时不使用`PTuning`. `pre_seq_len`为None/0, 

   样本

   ```python
       text = [
           '你好啊',
           '今天天气不错呢！'
       ]
       encode = tokenizer(text, return_tensors='pt', padding=True, add_special_tokens=False)
       print(encode['attention_mask'])
       
       """
   tensor([[0, 0, 1, 1, 1],
           [1, 1, 1, 1, 1]])    
       """
   ```

   `ChatGLMModel.forward`中处理`attention_mask`逻辑.  第一个if 不执行,第二个if 执行进入`get_masks`

   ```python
           if self.pre_seq_len is not None:
               if past_key_values is None:
                   past_key_values = self.get_prompt(batch_size=batch_size, device=input_ids.device,
                                                     dtype=inputs_embeds.dtype)
               if attention_mask is not None:
                   attention_mask = torch.cat([attention_mask.new_ones((batch_size, self.pre_seq_len)),
                                               attention_mask], dim=-1)
   
           if full_attention_mask is None:
               if (attention_mask is not None and not attention_mask.all()) or (past_key_values and seq_length != 1):
                   full_attention_mask = self.get_masks(input_ids, past_key_values, padding_mask=attention_mask)
                   
   ```

   调用`get_masks`后的结果

   ```python
       text = [
           '你好啊',
           '今天天气不错呢！'
       ]
       encode = tokenizer(text, return_tensors='pt', padding=True, add_special_tokens=False)
       print(encode)
       print('=' * 128)
       res = get_masks(encode['input_ids'], past_key_values=None, padding_mask=encode['attention_mask'])
       print(res)
       
   """    
   {'input_ids': tensor([[    0,     0, 36474, 54591, 55674],
           [53456, 33071, 32884, 55282, 31404]]), 'attention_mask': tensor([[0, 0, 1, 1, 1],
           [1, 1, 1, 1, 1]]), 'position_ids': tensor([[0, 0, 0, 1, 2],
           [0, 1, 2, 3, 4]])}
   ================================================================================================================================        
   tensor([[[[False, False, False, False, False],
             [False, False, False, False, False],
             [ True,  True, False,  True,  True],
             [ True,  True, False, False,  True],
             [ True,  True, False, False, False]]],
   
   
           [[[False,  True,  True,  True,  True],
             [False, False,  True,  True,  True],
             [False, False, False,  True,  True],
             [False, False, False, False,  True],
             [False, False, False, False, False]]]])
   """
   ```

   MHA的逻辑`CoreAttention.forward`, 下面执行else语句再反转attn

   ```python
       def forward(self, query_layer, key_layer, value_layer, attention_mask):
           pytorch_major_version = int(torch.__version__.split('.')[0])
           if pytorch_major_version >= 2:
               query_layer, key_layer, value_layer = [k.permute(1, 2, 0, 3) for k in [query_layer, key_layer, value_layer]]
               if attention_mask is None and query_layer.shape[2] == key_layer.shape[2]:
                   context_layer = torch.nn.functional.scaled_dot_product_attention(query_layer, key_layer, value_layer,
                                                                                    is_causal=True)
               else:
                   if attention_mask is not None:
                       attention_mask = ~attention_mask
                   context_layer = torch.nn.functional.scaled_dot_product_attention(query_layer, key_layer, value_layer,
                                                                                    attention_mask)
               context_layer = context_layer.permute(2, 0, 1, 3)
               new_context_layer_shape = context_layer.size()[:-2] + (self.hidden_size_per_partition,)
               context_layer = context_layer.reshape(*new_context_layer_shape)
   
    # 进入 torch.nn.functional.scaled_dot_product_attention的attn张这样
   """
   tensor([[[[ True,  True,  True,  True,  True],
             [ True,  True,  True,  True,  True],
             [False, False,  True, False, False],
             [False, False,  True,  True, False],
             [False, False,  True,  True,  True]]],
   
   
           [[[ True, False, False, False, False],
             [ True,  True, False, False, False],
             [ True,  True,  True, False, False],
             [ True,  True,  True,  True, False],
             [ True,  True,  True,  True,  True]]]])
   """
   ```

   可以看到第一个mask矩阵不符合预期. 预期应该张这样子.

   ```python
   """
   
   tensor([[[[False, False, False, False, False],
             [False, False, False, False, False],
             [False, False,  True, False, False],
             [False, False,  True,  True, False],
             [False, False,  True,  True,  True]]],
   
   
           [[[ True, False, False, False, False],
             [ True,  True, False, False, False],
             [ True,  True,  True, False, False],
             [ True,  True,  True,  True, False],
             [ True,  True,  True,  True,  True]]]])
   """
   ```

### Steps To Reproduce

对这个问题，我重新作了实现，在[modeling_chatglm.py的精简优化-2](https://github.com/diuzi/ChatWithAngel/tree/master/src/models/chatglm2)

### Environment

```markdown
- OS:
- Python:
- Transformers:
- PyTorch:
- CUDA Support (`python -c "import torch; print(torch.cuda.is_available())"`) :
```

**Comment by duzx16** [comment]
前两个 pad token 用不到所以 attention 是什么样的其实无所谓。之所以全部是 True 是因为如果是 False 的话，在不使用 `scaled_dot_product_attention` 的时候对应的 attention score全部是 `-inf`，会导致计算出来的 hidden state 变成 nan。

**Comment by diuzi** [comment]
多谢解惑
